src/interfaces/simulation.ts

```diff
@@ -496,20 +472,149 @@ export class SimulationInterface<T extends ChainId> extends ServiceInterface<T>
+      approvalTransactionId,
+      forkId,
+    };
+  }
+
+  private async getZapperZapInSimulationArgs({
+    depositArgs,
+    underlyingToken,
+  }: {
+    depositArgs: DepositArgs;
+    underlyingToken: UnderlyingToken;
+  }): Promise<{ simulateFn: (save: boolean) => Promise<TransactionOutcome>; forkId?: string }> {
+    if (!isEthereum(this.chainId)) {
+      throw new SdkError(`Zapper unsupported for chainId: ${this.chainId}`);
+    }
+
+    if (!depositArgs.options.slippage) {
+      throw new SdkError("slippage needs to be specified for a zap", SdkError.NO_SLIPPAGE);
+    }
+
+    const { needsApproving, approvalTransactionId, forkId } = await this.getZapperZapInApprovalData(depositArgs);
+
+    const simulateFn = (save: boolean): Promise<TransactionOutcome> => {
+      return this.zapIn({
+        depositArgs: {
+          ...depositArgs,
+          options: { ...depositArgs.options, forkId, save, root: approvalTransactionId },
+        },
+        underlyingToken,
+        skipGasEstimate: needsApproving,
+      });
+    };
+
+    return { simulateFn, forkId };
+  }
+
+  private async approve({ from, sellToken, amount, toVault, options }: DepositArgs): Promise<string> {
+    const tokenAbi = ["function approve(address spender, uint256 amount) returns (bool)"];
+    const signer = this.ctx.provider.write.getSigner(from);
+    const tokenContract = new Contract(sellToken, tokenAbi, signer);
+
+    const addressToApprove = this.getActionableAddress({ toVault });
+
+    const encodedInputData = tokenContract.interface.encodeFunctionData("approve", [addressToApprove, amount]);
+
+    const simulationResponse: SimulationResponse = await this.simulationExecutor.makeSimulationRequest(
+      from,
+      sellToken,
+      encodedInputData,
+      { ...options, save: true }
     );
+
+    return simulationResponse.simulation.id;
+  }
+
+  private async getApprovalData({
+    depositArgs,
+    signer,
+  }: {
+    depositArgs: DepositArgs;
+    signer: JsonRpcSigner;
+  }): Promise<ApprovalData> {
+    const depositNeedsApproving = await this.depositNeedsApproving({ depositArgs, signer });
+
+    if (!depositNeedsApproving) {
+      return {};
+    }
+
+    const approvalTransactionId = await this.approve(depositArgs);
+
+    const forkId = await this.simulationExecutor.createFork();
+
+    return { approvalTransactionId, forkId };
+  }
+
+  private getVaultContract({ toVault, signer }: { toVault: Address; signer: JsonRpcSigner }): VaultContractType {
+    const zapProtocol = this.yearn.services.zapper.getZapProtocol({ vault: toVault });
+
+    if (zapProtocol === ZapProtocol.PICKLE) {
+      return new PickleJarContract(toVault, signer);
+    }
+
+    return new YearnVaultContract(toVault, signer);
+  }
+
+  private async handleZapInSimulationDeposit({ zapInWith, vaultContract, depositArgs }: ZapInSimulationDepositArgs) {
+    if (zapInWith === "zapperZapIn") {
+      const underlyingToken = await this.getUnderlyingToken(vaultContract);
```

Comment:
Lets add a TODO here, to use a token.get(address) that returns Token

Reply:
Saw you added `token.findByAddress`. Should we use that rather than creating that `UnderlyingToken` interface?

Reply:
This does a bit more than getting the token, can we get all of this from a future `token.get(address)`?
```typescript
  private async getUnderlyingToken(vaultContract: VaultContractType): Promise<UnderlyingToken> {
    const [address, decimals, pricePerShare] = await Promise.all([
      vaultContract.token().catch(() => {
        throw new EthersError("failed to fetch token", EthersError.FAIL_TOKEN_FETCH);
      }),
      vaultContract.decimals().catch(() => {
        throw new EthersError("no decimals", EthersError.NO_DECIMALS);
      }),
      vaultContract.pricePerShare().catch(() => {
        throw new EthersError("no price per share", EthersError.NO_PRICE_PER_SHARE);
      }),
    ]);

    return { address, decimals, pricePerShare };
  }
```

Reply:
sorry got confused. 

Think all this data is available in the Vault interface, so how about using that one?

Reply:
It appears at the moment the vault interface (`vaults.getDynamic([toVault])`) only gives us the `amout`/`amountUsdc`

<img width="1248" alt="yearn_finance" src="https://user-images.githubusercontent.com/78794805/164950118-392d4ddc-1b90-4d84-8e1e-6db21f9f7633.png">

from the tokens interface, i.e. `tokens.findByAddress(sellToken)`

<img width="997" alt="yearn_finance" src="https://user-images.githubusercontent.com/78794805/164950174-74480b98-dbd9-4292-888a-276892148bc2.png">

Reply:
with `vaults.get([toVault])` we should obtain token, decimals and pricePerShare (in metadata)

Reply:
> with `vaults.get([toVault])` we should obtain token, decimals and pricePerShare (in metadata)

You're right, so I'm thinking just deleting `getUnderlyingToken` and refactor to use the `vaults.get([toVault])` instead of `vaults.getDynamic([toVault])`

<img width="1105" alt="yearn_finance" src="https://user-images.githubusercontent.com/78794805/165195262-ba1f8fae-0967-452a-b83c-f8cb4f52bb3f.png">

Reply:
Done with https://github.com/yearn/yearn-sdk/pull/280/commits/c062e73da0f8dea3201e782633c37296fa97fde0

Had to refactor a bit the tests